docs/source/python/dataset.rst

```diff
@@ -0,0 +1,393 @@
+.. Licensed to the Apache Software Foundation (ASF) under one
+.. or more contributor license agreements.  See the NOTICE file
+.. distributed with this work for additional information
+.. regarding copyright ownership.  The ASF licenses this file
+.. to you under the Apache License, Version 2.0 (the
+.. "License"); you may not use this file except in compliance
+.. with the License.  You may obtain a copy of the License at
+
+..   http://www.apache.org/licenses/LICENSE-2.0
+
+.. Unless required by applicable law or agreed to in writing,
+.. software distributed under the License is distributed on an
+.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+.. KIND, either express or implied.  See the License for the
+.. specific language governing permissions and limitations
+.. under the License.
+
+.. currentmodule:: pyarrow.dataset
+
+.. _dataset:
+
+Tabular Datasets
+================
+
+.. warning::
+
+    The ``pyarrow.dataset`` module is experimental (specifically the classes),
+    and a stable API is not yet guaranteed.
+
+The ``pyarrow.dataset`` module provides functionality to efficiently work with
+tabular, potentially larger than memory and multi-file, datasets:
+
+* A unified interface for different sources: supporting different sources
+  (files, database connection, ..), different file systems (local, cloud) and
+  different file formats (Parquet, CSV, JSON, Feather, ..)
+* Discovery of sources (crawling directories, handle directory-based partitioned
+  datasets, basic schema normalization, ..)
+* Optimized reading with pedicate pushdown (filtering rows), projection
+  (selecting columns), parallel reading or fine-grained managing of tasks.
+
+
+For those familiar with the existing :class:`pyarrow.parquet.ParquetDataset` for
+reading Parquet datasets: ``pyarrow.dataset``'s goal is similar but not specific
+to the Parquet format and not tied to Python: the same datasets API is exposed
+in the R bindings or Arrow. In addition ``pyarrow.dataset`` boasts improved
+perfomance and new features (e.g. filtering within files rather than only on
+partition keys).
+
+
+
+Reading Datasets
+----------------
+
+
+
+.. TODO Full blown example with NYC taxi data to show off, afterwards explain all parts:
+
+.. ipython:: python
```

Comment:
Hmm, I'm frankly not fond of "ipython" directives. The more we add, the slower building the docs becomes and the more it deters from editing/improving them.

Reply:
Yeah, I am happy to change this (I mostly used it because it was already used elsewhere). 

However, I think ideally we still check the code examples for correctness, where applicable. For example by using pytest doctests on them. This can be run separate as tests, and doesn't need to be part of doc building.